src/Extend/Frontend.php

```diff
@@ -92,6 +107,26 @@ private function registerRoutes(Container $container)
         }
     }
 
+    private function registerDocumentCallback(Container $container)
+    {
+        if (! isset($this->documentCallback)) {
+            return;
+        }
+
+        $container->resolving(
+            "flarum.$this->frontend.frontend",
+            function (HtmlDocumentFactory $view, Container $container) {
+                if ($this->documentCallback instanceof Closure) {
+                    $callback = $this->documentCallback;
+                } else {
+                    $callback = $container->make($this->documentCallback);
+                }
+
+                $view->addCallback($callback);
```

Comment:
Is it necessary to add a new collection of "callbacks" to the HtmlDocumentFactory? That's what "content" is for. We could do this instead:

```php
$content = new class implements ContentInterface {
    public $callback;
    public function populate(HtmlDocument $view, Request $request)
    {
        $this->callback($view, $request);
    }
});

$content->callback = $callback;

$view->add($content);
```

Reply:
Hmm, good point. What about getting rid of `ContentInterface` and making those classes invokable, too?

Reply:
Or we can keep the interface, and the existing listener class would implement that interface instead of being invokable.

Reply:
I like your second suggestion, as per my comment in the other thread :)

Reply:
I've done this now. I changed `ContentInterface` to use `__invoke` instead of `populate`, so that we can just drop in closures instead (we weren't type-hinting anyway).